Natural keys and natural_key_slug inspired by django-natural-keys (#3423)
nautobot/nautobot: nautobot/docs/development/natural-keys.md

# Natural Keys in Nautobot Models

+++ 2.0.0

Nautobot models derived from `BaseModel` automatically support the following [natural key](https://docs.djangoproject.com/en/3.2/topics/serialization/#natural-keys) APIs:

- Django's `instance.natural_key()` and `Model.objects.get_by_natural_key()` method APIs
- `instance.natural_key_slug` property inspired by the `django-natural-keys` project
- `Model.objects.get(natural_key_slug=...)` lookup filter inspired by the `django-natural-keys` project

## Using the Natural Key APIs

The `natural_key()` and `get_by_natural_key()` APIs are symmetric with one another:

```python
>>> DeviceType.objects.first().natural_key()
['Arista', 'Arista DeviceType 15']

>>> DeviceType.objects.get_by_natural_key("Arista", "Arista DeviceType 15")
<DeviceType: Arista DeviceType 15>
```

Similarly, the `natural_key_slug` and `get(natural_key_slug=...`) APIs are also symmetric:

```python
# Note that natural_key_slug is a property, not a method!
>>> DeviceType.objects.first().natural_key_slug
'Arista&Arista+DeviceType+15'

> >>> DeviceType.objects.get(natural_key_slug='Arista&Arista+DeviceType+15')

Reviewer: ```suggestion
>>> DeviceType.objects.get(natural_key_slug="Arista&Arista+DeviceType+15")
```

<DeviceType: Arista DeviceType 15>
```

!!! note

>     The `natural_key_slug` is designed to be suitable for use in URL patterns, such as an object detail endpoint `/app/model/<natural_key_slug>/`, though some caution must be taken when defining URL patterns containing such a slug, as it **can** contain the character `/` as a component.

Author: Holding on this comment until we reach a conclusion about inclusion or escaping of `/` characters (https://github.com/nautobot/nautobot/pull/3423/files#r1139300825)

## Implementing the Natural Key APIs

> In many model cases, Nautobot's default implementation of these APIs will suffice - as long as your model has any of the following, a default natural key will be automatically made available:

Reviewer: ```suggestion
In many model cases, Nautobot's default implementation of these APIs will suffice. As long as your model has any of the following, a default natural key will be automatically made available:
```

- One or more `UniqueConstraint` declarations
- Any `unique_together` declaration
- Any field (other than `id`) that is set as `unique=True`.

> There are a few special cases that will need special handling.

Reviewer: ```suggestion
There are a few special cases that will need special handling as described below.
```

### Self-Referential Natural Keys

An example of this can be seen with the `Location` model, where a given instance is only uniquely identified by its name **in combination with its parent**, which is another `Location`. Nautobot's default implementation would fall into an infinite recursion when trying to identify the Location's natural key fields, since they would be identified as  `("name", "parent__name", "parent__parent__name", "parent__parent__parent__name", ...)`.

> In a case like this, Nautobot is able to support _variadic_ natural keys, where the number and listing of natural keys may vary depending on the data of a given instance. To make this work, you will need to override two natural key related APIs on your model, `natural_key_field_lookups` and `natural_key_args_to_kwargs`, as follows:

Reviewer: ```suggestion
In a case like this, Nautobot is able to support _variadic_ natural keys, where the number and listing of natural keys may vary depending on the data of a given instance. To make this work, you will need to override two APIs related to natural keys on your model (`natural_key_field_lookups` and `natural_key_args_to_kwargs`) as follows:
```

```python
class Location(TreeModel):

    class Meta:
        unique_together = [["parent", "name"]]

    @classproperty
    def natural_key_field_lookups(cls):
        """
        Due to the recursive nature of Location's natural key, we need a custom implementation of this property.

        This returns a set of natural key lookups based on the current maximum depth of the Location tree.
        For example if the tree is 2 layers deep, it will return ["name", "parent__name", "parent__parent__name"].

        Without this custom implementation, the generic `natural_key_field_lookups` would recurse infinitely.
        """
        lookups = []
        name = "name"
        for _ in range(cls.objects.max_tree_depth() + 1):
            lookups.append(name)
            name = f"parent__{name}"
        return lookups

    @classmethod

>     def natural_key_args_to_kwargs(cls, args):
>         """Handle the possibility that more recursive "parent" lookups were specified than we initially expected."""
>         args = list(args)

Reviewer: ```suggestion
    def natural_key_args_to_kwargs(cls, *args):
        """Handle the possibility that more recursive "parent" lookups were specified than we initially expected."""
```

Author: The reason for the explicit cast to `list()` is because `args` may be an immutable tuple, so without casting it to a list we cannot append to it.

I can change the signature from `(args)` to `(*args)` if you think that would be preferable, but in practice I found that most of the time I already have a list or tuple of args to pass in, rather than distinct individual parameters, so I'd have to explode that list/tuple when calling the function, i.e. instead of (current pattern):

```python
args = ["hello", "world"]
kwargs = model.natural_key_args_to_kwargs(args)
```

I'd have to do:

```python
args = ["hello", "world"]
kwargs = model.natural_key_args_to_kwargs(*args)
```

which just felt clunky to me.

Reviewer: Ahh, I see. I also realize I was incorrect in my thought that the unpacked list would be passed through as such, and it is not. It comes through as a `tuple`.

        natural_key_field_lookups = list(cls.natural_key_field_lookups)
        while len(args) < len(natural_key_field_lookups):
            args.append(None)
        while len(args) > len(natural_key_field_lookups):
            natural_key_field_lookups.append(f"parent__{natural_key_field_lookups[-1]}")
        return dict(zip(natural_key_field_lookups, args))
```

### Natural Keys Referencing a Different Self-Referential Model

> Relatedly, if you have a model whose natural keys include a ForeignKey to a model like `Location` with self-referential and variadic natural keys, for the related model to be handled properly, you must always ensure that the related field is the **last** such field in your model's uniqueness constraint or `natural_key_field_names` declaration. Instead of:

Reviewer: ```suggestion
Similarly, if you have a model whose natural keys include a `ForeignKey` to a model like `Location` with self-referential and variadic natural keys, for the related model to be handled properly, you must always ensure that the related field is the **last** such field in your model's uniqueness constraint or `natural_key_field_names` declaration.

Instead of this:
```

```python
class VLANGroup(BaseModel):
    class Meta:
        unique_together = [["location", "name"]]   # wrong, nested location natural key cannot be variadic
```

> do:

Reviewer: ```suggestion
Do this:
```
